google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/project/DefaultAppEngineProjectService.java

```diff
@@ -197,4 +212,72 @@ public AppEngineStandardRuntime getAppEngineStandardDeclaredRuntime(
     }
   }
 
+  @Override
+  public Optional<String> getServiceNameFromAppYaml(@NotNull String appYamlPathString) {
+    return getValueFromAppYaml(appYamlPathString, SERVICE_TAG_YAML);
+  }
+
+  @Override
+  public FlexibleRuntime getFlexibleRuntimeFromAppYaml(@NotNull String appYamlPathString) {
+    return getValueFromAppYaml(appYamlPathString, RUNTIME_TAG_YAML).map(runtimeValue -> {
+      if (runtimeValue.equals(JAVA_RUNTIME)) {
+        return FlexibleRuntime.JAVA;
+      }
+      if (runtimeValue.equals(CUSTOM_RUNTIME)) {
+        return FlexibleRuntime.CUSTOM;
+      }
+      return FlexibleRuntime.UNKNOWN;
+    }).orElse(FlexibleRuntime.UNKNOWN);
+  }
+
+  private Optional<String> getValueFromAppYaml(@NotNull String appYamlPathString,
+      @NotNull String key) {
+    Path appYamlPath;
+    try {
+      appYamlPath = Paths.get(appYamlPathString);
+    } catch (InvalidPathException ipe) {
+      return Optional.empty();
+    }
+
+    try {
+      return Files.readAllLines(appYamlPath).stream()
```

Comment:
this bit is important.  I don't think we're successfully parsing all valid configs with this code.